nsdb-cluster/src/main/scala/io/radicalbit/nsdb/cluster/coordinator/MetadataCoordinator.scala

```diff
@@ -442,43 +425,49 @@ class MetadataCoordinator(clusterListener: ActorRef,
           timestamp,
           MetadataCoordinator.notEnoughReplicasErrorMessage(clusterAliveMembers.size, replicationFactor))
       else {
-        val chain: Future[GetWriteLocationsResponse] = (metadataCache ? GetLocationsFromCache(db, namespace, metric))
-          .flatMap {
-            case LocationsCached(_, _, _, values) =>
-              values.filter(v => v.from <= timestamp && v.to >= timestamp) match {
-                case Nil =>
-                  for {
-                    shardInterval <- getShardInterval(db, namespace, metric)
-                    nodeMetrics   <- (clusterListener ? GetNodeMetrics).mapTo[NodeMetricsGot]
-                    addLocationResult <- {
-                      val start = getShardStartInstant(timestamp, shardInterval)
-                      val end   = getShardEndInstant(start, shardInterval)
-
-                      val nodes =
-                        if (nodeMetrics.nodeMetrics.nonEmpty)
-                          writeNodesSelectionLogic.selectWriteNodes(nodeMetrics.nodeMetrics, replicationFactor)
-                        else {
-                          Random.shuffle(clusterAliveMembers.toSeq).take(replicationFactor).map(createNodeName)
+        val currentTime = System.currentTimeMillis()
+        getMetricInfo(db, namespace, metric) { metricInfo =>
+          val retention = metricInfo.retention
+          if (retention > 0 && timestamp < currentTime - retention || timestamp > currentTime + retention)
+            Future(GetWriteLocationsBeyondRetention(db, namespace, metric, timestamp, metricInfo.retention))
+          else {
+            (metadataCache ? GetLocationsFromCache(db, namespace, metric))
+              .flatMap {
+                case LocationsCached(_, _, _, values) =>
+                  values.filter(v => v.from <= timestamp && v.to >= timestamp) match {
```

Comment:
if `v` is an object that we handle, it might have a method `.contains(timestamp: Long)`

Reply:
yes, v is an object of ours (a `Location`). And yes, we should totally add a `contains` method.
Thanks.